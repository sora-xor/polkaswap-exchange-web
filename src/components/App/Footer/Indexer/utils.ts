import { IndexerType } from '@soramitsu/soraneo-wallet-web/lib/consts';

export function getIndexerName(type: IndexerType): string {
  return type.charAt(0).toUpperCase() + type.slice(1).toLowerCase();
}
