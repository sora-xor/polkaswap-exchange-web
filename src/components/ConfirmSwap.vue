<template>
  <dialog-base
    :visible.sync="isVisible"
    :title="t('swap.confirmSwap')"
    custom-class="dialog--confirm-swap"
  >
    <div class="tokens">
      <div class="tokens-info-container">
        <span class="token-value">{{ formattedFromValue }}</span>
        <div v-if="tokenFrom" class="token">
          <token-logo :token="tokenFrom" />
          {{ tokenFrom.symbol }}
        </div>
      </div>
      <s-icon class="icon-divider" name="arrow-bottom-rounded" size="medium" />
      <div class="tokens-info-container">
        <span class="token-value">{{ formattedToValue }}</span>
        <div v-if="tokenTo" class="token">
          <token-logo :token="tokenTo" />
          {{ tokenTo.symbol }}
        </div>
      </div>
    </div>
    <p :class="isExchangeB ? 'transaction-message' : 'transaction-message transaction-message--min-received'" v-html="t('swap.swapOutputMessage', { transactionValue : `<span class='transaction-number'>${ formattedMinMaxReceived }</span>` })" />
    <s-divider />
    <swap-info :show-price="true" />
    <swap-info :show-tooltips="false" />
    <template #footer>
      <s-button
        type="primary"
        :disabled="loading"
        @click="handleConfirmSwap"
      >
        {{ t('exchange.confirm') }}
      </s-button>
    </template>
  </dialog-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Getter } from 'vuex-class'
import { api } from '@soramitsu/soraneo-wallet-web'
import { CodecString, AccountAsset } from '@sora-substrate/util'

import TransactionMixin from '@/components/mixins/TransactionMixin'
import DialogMixin from '@/components/mixins/DialogMixin'
import NumberFormatterMixin from '@/components/mixins/NumberFormatterMixin'
import DialogBase from '@/components/DialogBase.vue'
import { lazyComponent } from '@/router'
import { Components } from '@/consts'

const namespace = 'swap'

@Component({
  components: {
    DialogBase,
    SwapInfo: lazyComponent(Components.SwapInfo),
    TokenLogo: lazyComponent(Components.TokenLogo)
  }
})
export default class ConfirmSwap extends Mixins(TransactionMixin, DialogMixin, NumberFormatterMixin) {
  @Getter('tokenFrom', { namespace }) tokenFrom!: AccountAsset
  @Getter('tokenTo', { namespace }) tokenTo!: AccountAsset
  @Getter('fromValue', { namespace }) fromValue!: string
  @Getter('toValue', { namespace }) toValue!: string
  @Getter('minMaxReceived', { namespace }) minMaxReceived!: CodecString
  @Getter('isExchangeB', { namespace }) isExchangeB!: boolean

  @Getter slippageTolerance!: number

  @Prop({ default: false, type: Boolean }) readonly isInsufficientBalance!: boolean

  get formattedFromValue (): string {
    return this.getFPNumber(this.fromValue, this.tokenFrom?.decimals).format()
  }

  get formattedToValue (): string {
    return this.getFPNumber(this.toValue, this.tokenTo?.decimals).format()
  }

  get formattedMinMaxReceived (): string {
    const decimals = (this.isExchangeB ? this.tokenFrom : this.tokenTo)?.decimals
    return this.formatCodecNumber(this.minMaxReceived, decimals)
  }

  async handleConfirmSwap (): Promise<void> {
    await this.$emit('checkConfirm')
    if (this.isInsufficientBalance) {
      this.$alert(this.t('exchange.insufficientBalance', { tokenSymbol: this.tokenFrom ? this.tokenFrom.symbol : '' }), { title: this.t('errorText') })
      this.$emit('confirm')
    } else {
      try {
        await this.withNotifications(
          async () => await api.swap(
            this.tokenFrom.address,
            this.tokenTo.address,
            this.fromValue,
            this.toValue,
            this.slippageTolerance,
            this.isExchangeB
          )
        )
        this.$emit('confirm', true)
      } catch (error) {
        this.$emit('confirm')
      }
    }
    this.isVisible = false
  }
}
</script>

<style lang="scss">
.dialog--confirm-swap {
  .transaction-message {
    .min-received-label {
      display: none;
    }
    &--min-received {
      .min-received-label {
        display: inline;
      }
    }
  }
  .transaction-number {
    color: var(--s-color-base-content-primary);
    font-feature-settings: $s-font-feature-settings-common;
    @include font-weight(600);
    word-break: break-all;
  }
}
</style>

<style lang="scss" scoped>
.tokens {
  display: flex;
  flex-direction: column;
  font-size: var(--s-heading2-font-size);
  line-height: $s-line-height-small;
  &-info-container {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
}
.token {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  white-space: nowrap;
  &-value {
    margin-right: $inner-spacing-medium;
  }
  &-logo {
    display: block;
    margin-right: $inner-spacing-medium;
    flex-shrink: 0;
  }
}
.transaction-message {
  margin-top: $inner-spacing-mini;
  color: var(--s-color-base-content-tertiary);
  line-height: $s-line-height-big;
}
@include vertical-divider;
@include vertical-divider('el-divider');
</style>
