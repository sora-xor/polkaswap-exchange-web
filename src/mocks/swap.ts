export const tokens: any = {
  XOR: {
    name: 'Sora',
    logo: '',
    symbol: 'XOR',
    address: '1f9840a85d5af5bf1d1762f925bdaddc4201f984',
    balance: 10000,
    price: 0.0025
  },
  KSM: {
    name: 'Kusama',
    logo: '',
    symbol: 'KSM',
    address: '34916349d43f65bccca11ff53a8e0382a1a594a7',
    balance: 0,
    price: 0.0055
  },
  ETH: {
    name: 'Ether',
    logo: '',
    symbol: 'ETH',
    address: '8adaca8ea8192656a15c88797e04c8771c4576b3',
    balance: 20000.45,
    price: 0.0099
  }
}
export const slippageTolerance = 0.5
export const liquidityProviderFee = 0.3
