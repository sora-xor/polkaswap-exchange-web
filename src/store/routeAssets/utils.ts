import { FPNumber, Operation } from '@sora-substrate/util/build';
import { Asset, AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '@soramitsu/soraneo-wallet-web';
import { FiatPriceObject } from '@soramitsu/soraneo-wallet-web/lib/services/subquery/types';

import { Recipient } from './types';

import type { WhitelistArrayItem } from '@sora-substrate/util/build/assets/types';

export default {
  validate(recipient: Recipient) {
    return (
      this.wallet(recipient.wallet) &&
      this.asset(recipient.asset) &&
      this.usd(recipient.usd) &&
      this.amount(recipient.amount)
    );
  },

  name(name: Nullable<string>) {
    return name;
  },

  wallet(wallet: string) {
    return api.validateAddress(wallet);
  },

  asset(asset: Asset | AccountAsset) {
    return asset;
  },

  usd(usd: Nullable<number>) {
    return usd && !isNaN(usd);
  },

  amount(amount: Nullable<number>) {
    return amount && !isNaN(amount) && Number.isFinite(amount);
  },
};

export function getTokenEquivalent(
  priceObject: FiatPriceObject,
  asset: Asset | AccountAsset | WhitelistArrayItem,
  usd: number | string
): FPNumber {
  return new FPNumber(usd).div(FPNumber.fromCodecValue(priceObject[asset.address], asset.decimals));
}

export function getAssetUSDPrice(asset, fiatPriceObject) {
  if (!asset) return null;
  return FPNumber.fromCodecValue(fiatPriceObject[asset.address], asset.decimals);
}
