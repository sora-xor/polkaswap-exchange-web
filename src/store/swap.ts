import map from 'lodash/fp/map'
import flatMap from 'lodash/fp/flatMap'
import fromPairs from 'lodash/fp/fromPairs'
import flow from 'lodash/fp/flow'
import concat from 'lodash/fp/concat'
import { api } from '@soramitsu/soraneo-wallet-web'
import { KnownAssets, KnownSymbols, Asset, AccountAsset, CodecString } from '@sora-substrate/util'

import { ZeroStringValue } from '@/consts'

const types = flow(
  flatMap(x => [x + '_REQUEST', x + '_SUCCESS', x + '_FAILURE']),
  concat([
    'SET_FROM_VALUE',
    'SET_TO_VALUE',
    'SET_TOKEN_FROM_PRICE',
    'SET_MIN_MAX_RECEIVED',
    'SET_EXCHANGE_B',
    'SET_LIQUIDITY_PROVIDER_FEE',
    'SET_NETWORK_FEE',
    'GET_SWAP_CONFIRM'
  ]),
  map(x => [x, x]),
  fromPairs
)([
  'GET_TOKEN_XOR',
  'GET_TOKEN_FROM',
  'GET_TOKEN_TO'
])

interface SwapState {
  tokenXOR: Asset | AccountAsset | null;
  tokenFrom: Asset | AccountAsset | null;
  tokenTo: Asset | AccountAsset | null;
  fromValue: string;
  toValue: string;
  isTokenFromPrice: boolean;
  minMaxReceived: CodecString;
  isExchangeB: boolean;
  liquidityProviderFee: CodecString;
  networkFee: CodecString;
}

function initialState (): SwapState {
  return {
    tokenXOR: null,
    tokenFrom: null,
    tokenTo: null,
    fromValue: '',
    toValue: '',
    isTokenFromPrice: true,
    minMaxReceived: '',
    isExchangeB: false,
    liquidityProviderFee: '',
    networkFee: ''
  }
}

const state = initialState()

const getters = {
  tokenXOR (state: SwapState) {
    return state.tokenXOR
  },
  tokenFrom (state: SwapState) {
    return state.tokenFrom
  },
  tokenTo (state: SwapState) {
    return state.tokenTo
  },
  fromValue (state: SwapState) {
    return state.fromValue
  },
  toValue (state: SwapState) {
    return state.toValue
  },
  isTokenFromPrice (state: SwapState) {
    return state.isTokenFromPrice
  },
  minMaxReceived (state: SwapState) {
    return state.minMaxReceived
  },
  isExchangeB (state: SwapState) {
    return state.isExchangeB
  },
  liquidityProviderFee (state: SwapState) {
    return state.liquidityProviderFee
  },
  networkFee (state: SwapState) {
    return state.networkFee
  }
}

const mutations = {
  [types.GET_TOKEN_XOR_REQUEST] (state: SwapState) {
    state.tokenXOR = null
  },
  [types.GET_TOKEN_XOR_SUCCESS] (state: SwapState, token: Asset | AccountAsset | null) {
    state.tokenXOR = token
  },
  [types.GET_TOKEN_XOR_FAILURE] (state: SwapState) {
    state.tokenXOR = null
  },
  [types.GET_TOKEN_FROM_REQUEST] (state: SwapState) {
    state.tokenFrom = null
  },
  [types.GET_TOKEN_FROM_SUCCESS] (state: SwapState, token: Asset | AccountAsset | null) {
    state.tokenFrom = token
  },
  [types.GET_TOKEN_FROM_FAILURE] (state: SwapState) {
    state.tokenFrom = null
  },
  [types.GET_TOKEN_TO_REQUEST] (state: SwapState) {
    state.tokenTo = null
  },
  [types.GET_TOKEN_TO_SUCCESS] (state: SwapState, token: Asset | AccountAsset | null) {
    state.tokenTo = token
  },
  [types.GET_TOKEN_TO_FAILURE] (state: SwapState) {
    state.tokenTo = null
  },
  [types.SET_FROM_VALUE] (state: SwapState, fromValue: string) {
    state.fromValue = fromValue
  },
  [types.SET_TO_VALUE] (state: SwapState, toValue: string) {
    state.toValue = toValue
  },
  [types.SET_TOKEN_FROM_PRICE] (state: SwapState, isTokenFromPrice: boolean) {
    state.isTokenFromPrice = isTokenFromPrice
  },
  [types.SET_MIN_MAX_RECEIVED] (state: SwapState, minMaxReceived: CodecString) {
    state.minMaxReceived = minMaxReceived
  },
  [types.SET_EXCHANGE_B] (state: SwapState, isExchangeB: boolean) {
    state.isExchangeB = isExchangeB
  },
  [types.SET_LIQUIDITY_PROVIDER_FEE] (state: SwapState, liquidityProviderFee: CodecString) {
    state.liquidityProviderFee = liquidityProviderFee
  },
  [types.SET_NETWORK_FEE] (state: SwapState, networkFee: CodecString) {
    state.networkFee = networkFee
  }
}

const actions = {
  async getTokenXOR ({ commit }) {
    const token = KnownAssets.get(KnownSymbols.XOR)
    commit(types.GET_TOKEN_XOR_REQUEST)
    try {
      if (token) {
        const tokenFrom = await api.accountAssets.find(asset => asset.address === token.address)
        if (tokenFrom) {
          commit(types.GET_TOKEN_XOR_SUCCESS, { ...tokenFrom })
        }
      } else {
        throw new Error(`There is no ${KnownSymbols.XOR} asset`)
      }
    } catch (error) {
      commit(types.GET_TOKEN_XOR_FAILURE)
      throw error
    }
  },
  async setTokenFrom ({ commit, rootGetters }, payload) {
    const token = KnownAssets.get(payload.tokenSymbol) || rootGetters['assets/assets'].find(item => item.symbol === payload.tokenSymbol)
    if (payload.isWalletConnected) {
      commit(types.GET_TOKEN_FROM_REQUEST)
      if (payload.updatedAsset) {
        commit(types.GET_TOKEN_FROM_SUCCESS, payload.updatedAsset)
        return
      }
      try {
        if (token) {
          let tokenFrom = await api.accountAssets.find(asset => asset.address === token.address)
          if (!tokenFrom) {
            tokenFrom = { ...token, balance: ZeroStringValue }
          }
          commit(types.GET_TOKEN_FROM_SUCCESS, tokenFrom)
        } else {
          throw new Error(`There is no ${payload.tokenSymbol} asset`)
        }
      } catch (error) {
        commit(types.GET_TOKEN_FROM_FAILURE)
        throw error
      }
    } else {
      commit(types.GET_TOKEN_FROM_SUCCESS, token)
    }
  },
  async setTokenTo ({ commit, rootGetters }, payload) {
    const token = KnownAssets.get(payload.tokenSymbol) || rootGetters['assets/assets'].find(item => item.symbol === payload.tokenSymbol)
    if (payload.isWalletConnected) {
      commit(types.GET_TOKEN_TO_REQUEST)
      if (payload.updatedAsset) {
        commit(types.GET_TOKEN_TO_SUCCESS, payload.updatedAsset)
        return
      }
      if (!payload.tokenSymbol) {
        return
      }
      try {
        if (token) {
          let tokenTo = await api.accountAssets.find(asset => asset.address === token.address)
          if (!tokenTo) {
            tokenTo = { ...token, balance: ZeroStringValue }
          }
          commit(types.GET_TOKEN_TO_SUCCESS, tokenTo)
        } else {
          throw new Error(`There is no ${payload.tokenSymbol} asset`)
        }
      } catch (error) {
        commit(types.GET_TOKEN_TO_FAILURE)
        throw error
      }
    } else {
      commit(types.GET_TOKEN_TO_SUCCESS, token)
    }
  },
  setFromValue ({ commit }, fromValue: string | number) {
    commit(types.SET_FROM_VALUE, fromValue)
  },
  setToValue ({ commit }, toValue: string | number) {
    commit(types.SET_TO_VALUE, toValue)
  },
  setTokenFromPrice ({ commit }, isTokenFromPrice: boolean) {
    commit(types.SET_TOKEN_FROM_PRICE, isTokenFromPrice)
  },
  setMinMaxReceived ({ commit }, minMaxReceived) {
    commit(types.SET_MIN_MAX_RECEIVED, minMaxReceived)
  },
  setExchangeB ({ commit }, flag: boolean) {
    commit(types.SET_EXCHANGE_B, flag)
  },
  setLiquidityProviderFee ({ commit }, liquidityProviderFee: string) {
    commit(types.SET_LIQUIDITY_PROVIDER_FEE, liquidityProviderFee)
  },
  setNetworkFee ({ commit }, networkFee: string) {
    commit(types.SET_NETWORK_FEE, networkFee)
  }
}

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
}
