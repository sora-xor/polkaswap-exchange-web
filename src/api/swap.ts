import { slippageTolerance, liquidityProviderFee } from '@/mocks/swap'

const getSlippageTolerance = () => {
  return Promise.resolve(slippageTolerance)
}

const getLiquidityProviderFee = () => {
  return Promise.resolve(liquidityProviderFee)
}

export default {
  getSlippageTolerance,
  getLiquidityProviderFee
}
