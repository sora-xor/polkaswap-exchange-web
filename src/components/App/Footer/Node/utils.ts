import { countryCodeEmoji } from 'country-code-emoji';

import store from '@/store';

export function formatLocation(code: string): Nullable<{ flag: string; name?: string }> {
  try {
    const isoCode = code.toUpperCase();
    const flag = countryCodeEmoji(isoCode);
    const location = { flag, name: '' };
    const displayRegions = store.state.settings.displayRegions;
    if (!displayRegions) {
      return location;
    }
    const name = displayRegions.of(isoCode);
    if (name) {
      location.name = name;
    }
    return location;
  } catch (error) {
    console.warn('Unsupported format of location in env.json', error);
    return null;
  }
}
