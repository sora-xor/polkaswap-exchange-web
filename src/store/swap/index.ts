import { defineModule } from 'direct-vuex';

import { localActionContext, localGetterContext } from '@/store';
import { Module } from '@/store/consts';

import actions from './actions';
import getters from './getters';
import mutations from './mutations';
import state from './state';

const swap = defineModule({
  namespaced: true,
  state,
  mutations,
  getters,
  actions,
});

const swapGetterContext = (args: [any, any, any, any]) => localGetterContext(args, Module.Swap, swap);
const swapActionContext = (context: any) => localActionContext(context, Module.Swap, swap);

export { swapGetterContext, swapActionContext };
export default swap;
