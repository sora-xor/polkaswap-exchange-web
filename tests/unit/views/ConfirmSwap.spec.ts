import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import SoramitsuElements from '@soramitsu/soramitsu-js-ui'
import ConfirmSwap from '@/components/ConfirmSwap.vue'
import { tokens } from '@/mocks/swap'
import { TranslationMock } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(SoramitsuElements)

describe('ConfirmSwap.vue', () => {
  let actions
  let getters
  let store

  beforeEach(() => {
    TranslationMock(ConfirmSwap)

    actions = {
      setSwapConfirm: jest.fn()
    }

    getters = {
      tokenFrom: () => tokens.XOR,
      tokenTo: () => tokens.KSM,
      fromValue: () => 150.123654,
      toValue: () => 0
    }

    store = new Vuex.Store({
      actions,
      getters
    })
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(ConfirmSwap, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
