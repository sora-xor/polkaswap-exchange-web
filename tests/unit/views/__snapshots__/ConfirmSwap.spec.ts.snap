// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmSwap.vue should renders correctly 1`] = `
<s-dialog-stub
  class="el-dialog--swap-confirm"
  closeonclickmodal="true"
  closeonesc="true"
  customclass=""
  lockscroll="true"
  modal="true"
  modalappendtobody="true"
  showclose="true"
  title=""
  top="auto"
  width="496px"
>
  <div
    class="tokens"
  >
    <div
      class="tokens-info-container"
    >
      <span
        class="token-value"
      >
        150.1237
      </span>
       
      <s-icon-stub
        name="arrow-bottom-rounded"
      />
       
      <span
        class="token-value"
      >
        0.0000
      </span>
    </div>
     
    <div
      class="tokens-info-container"
    >
      <div
        class="token"
      >
        <span
          class="token-logo token-logo--xor"
        />
        
        XOR
      
      </div>
       
      <div
        class="token"
      >
        <span
          class="token-logo token-logo--ksm"
        />
        
        KSM
      
      </div>
    </div>
  </div>
   
  <p
    class="transaction-message"
  />
   
  <s-divider-stub
    contentposition="center"
    direction="horizontal"
    type="secondary"
  />
   
  <swap-info-stub
    showprice="true"
  />
   
  <swap-info-stub />
</s-dialog-stub>
`;
